packages/@react-spectrum/s2/src/NotificationBadge.tsx

```diff
@@ -0,0 +1,146 @@
+import {filterDOMProps} from '@react-aria/utils';
+import {getAllowedOverrides, StyleProps} from './style-utils' with {type: 'macro'};
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import {NumberFormatter} from '@internationalized/number';
+import React, {createContext, forwardRef} from 'react';
+import {style} from '../style' with {type: 'macro'};
+import {useDOMRef} from '@react-spectrum/utils';
+import {useLocale, useLocalizedStringFormatter} from '@react-aria/i18n';
+import {useSpectrumContextProps} from './useSpectrumContextProps';
+
+export interface NotificationBadgeStyleProps {
+  /**
+   * The size of the notification badge.
+   *
+   * @default 'S'
+   */
+  size?: 'S' | 'M' | 'L' | 'XL'
+}
+
+export interface NotificationBadgeProps extends DOMProps, AriaLabelingProps, StyleProps, NotificationBadgeStyleProps, SlotProps {
+  /**
+   * The value to be displayed in the notification badge.
+   */
+  value?: number
+}
+
+export const NotificationBadgeContext = createContext<ContextValue<Partial<NotificationBadgeProps>, DOMRefValue<HTMLDivElement>>>(null);
+
+const badge = style({
+  display: 'flex',
+  font: 'control',
+  fontSize: {
+    size: {
+      S: 'ui-xs',
+      M: 'ui-xs',
+      L: 'ui-sm',
+      XL: 'ui'
+    }
+  },
+  justifyContent: 'center',
+  alignItems: 'center',
+  backgroundColor: 'accent',
+  height: {
+    size: {
+      S: {
+        default: 12,
+        isEmpty: 8
+      },
+      M: {
+        default: '[14px]',
+        isEmpty: 8
+      },
+      L: {
+        default: 16,
+        isEmpty: '[10px]'
+      },
+      XL: {
+        default: 18,
+        isEmpty: '[10px]'
+      }
+    }
+  },
+  aspectRatio: {
+    isEmpty: 'square',
+    isSingleDigit: 'square'
+  },
+  width: 'fit',
+  paddingX: {
+    isDoubleDigit: 'edge-to-text'
+  },
+  borderRadius: 'pill'
+}, getAllowedOverrides());
+
+/**
+ * Notification badges are used to indicate new or pending activity .
+ */
+export const NotificationBadge = forwardRef(function Badge(props: NotificationBadgeProps, ref: DOMRef<HTMLDivElement>) {
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-spectrum/s2');
+  [props, ref] = useSpectrumContextProps(props, ref, NotificationBadgeContext);
+  let {
+    size = 'S',
+    value,
+    ...otherProps
+  } = props;
+  let domRef = useDOMRef(ref);
+  let {locale} = useLocale();
+
+  if (value && value < 0) {
+    throw new Error('Value cannot be negative');
+  }
+
+  let truncatedValue: number | undefined = undefined;
+  if (value && value > 99) {
+    truncatedValue = 99;
+  }
+
+  let formattedValue = '';
+  if (truncatedValue) {
+    formattedValue = new NumberFormatter(locale).format(truncatedValue);
+  } else if (value) {
+    formattedValue = new NumberFormatter(locale).format(value);
+  }
+
+  let length = formattedValue.length;
+  let isEmpty = false;
+  let isSingleDigit = false;
+  let isDoubleDigit = false;
+  let isMaxDigit = false;
+  if (!value) {
```

Comment:
```suggestion
  if (value == undefined) {
```

Reply:
good habit to be in since 0 is falsy and everything else is truth
could also use isNaN, though search the codebase for which one to use, there's two of them